### PR TITLE
DAOS-4560 container: close container by IV

### PR DESCRIPTION
**Pull request by wangdi1**
1. Close container by IV so to unify the container
open/close.
2. close the container during fault injection
process.
3. shrink the iv tree width, since we already support
cart group update to recalculate cart tree, then
skip the failure nodes.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/container/container_iv.c [hunk 1]
```diff
@@ -383,6 +383,11 @@ cont_iv_ent_update(struct ds_iv_entry *entry, struct ds_iv_key *key,
 		/* If the uuid is NULL, then it will delete(invalidate) all
 		 * entries (containers under the pool) of the tree.
 		 */
+		if (entry->iv_class->iv_class_id == IV_CONT_CAPA) {
+			rc = ds_cont_tgt_close(civ_key->cont_uuid);
+			if (rc)
+				return rc;
+		}
```
Comment (liw) [inline src/container/container_iv.c:390]: Should these line be added before the comment from line 383 to 385? The comment seems to describe line 391 and downward.

File: src/container/srv_container.c [hunk 2]
```diff
@@ -1133,41 +1133,25 @@ static int
...
-	rc = out->tco_rc;
-	if (rc != 0) {
-		D_ERROR(DF_CONT": failed to close %d targets\n",
-			DP_CONT(svc->cs_pool_uuid, NULL), rc);
-		rc = -DER_IO;
+	/* update container capa to IV */
+	for (i = 0; i < nrecs; i++) {
+		rc = cont_iv_capability_invalidate(
+				svc->cs_pool->sp_iv_ns,
+			        recs[i].tcr_hdl, CRT_IV_SYNC_EAGER);
```
Comment (liw) [inline src/container/srv_container.c:1148]: So we lose the batching of container handle closes? Is there any similar batching ability from IV?

Comment (wangdi1) [reply src/container/srv_container.c:1148]: yes, indeed. I will try to pack multiple container handle in one IV req in the following patch.

File: src/container/container_iv.c (hunk 1 above)
Comment (liw) [reply src/container/container_iv.c:513]: This comment issue is here.

File: src/container/container_iv.c [hunk 3]
```diff
@@ -787,6 +792,7 @@ struct iv_capa_ult_arg {
 	uuid_t		pool_uuid;
 	uuid_t		cont_uuid;
 	uuid_t		cont_hdl_uuid;
+	bool		invalid_current;
```
Comment (liw) [inline src/container/container_iv.c:795]: Shortening "invalidate" to "invalid" is potentially confusing for a boolean field.

**Review by liw — APPROVED** [review]

**Review by liw — DISMISSED** [review]

**Review by jolivier23 — DISMISSED** [review]

**Review by sydneyvanda — CHANGES_REQUESTED** [review]
I don't see results for daos IO tests being run, specifically for IO38: io_capa_iv_fetch. Needs to be re-ran I think to make sure that this PR fixes the original issue (I think I/O tests were not run due to DAOS-5783 possible)

**Comment by wangdi1** [comment]
> I don't see results for daos IO tests being run, specifically for IO38: io_capa_iv_fetch. Needs to be re-ran I think to make sure that this PR fixes the original issue (I think I/O tests were not run due to DAOS-5783 possible)

ok, let me retrigger the test

**Review by kccain — COMMENTED** [review]
Just a couple of non-binding questions/comments. Mostly was trying to understand the existing and new structure.

File: src/container/srv_container.c [hunk 4]
```diff
@@ -1192,41 +1199,25 @@ static int
...
 		 struct cont_tgt_close_rec recs[], int nrecs)
 {
-	struct cont_tgt_close_in       *in;
-	struct cont_tgt_close_out      *out;
-	crt_rpc_t		       *rpc;
-	int				rc;
+	int	i;
+	int	rc = 0;
 
 	D_DEBUG(DF_DSMS, DF_CONT": bcasting: recs[0].hdl="DF_UUID
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/srv_container.c:1205]: ah, sure.

File: src/container/container_iv.c [hunk 5]
```diff
@@ -787,6 +791,7 @@ struct iv_capa_ult_arg {
 	uuid_t		pool_uuid;
 	uuid_t		cont_uuid;
 	uuid_t		cont_hdl_uuid;
+	bool		invalidate_current;
```
[the remark this answers is not in the document]
Comment (wangdi1) [reply src/container/container_iv.c:794]: sure, I will update if the patch needs refresh.

**Review by kccain — APPROVED** [review]

**Review by sydneyvanda — DISMISSED** [review]

**Comment by wangdi1** [comment]
Fix one issue found during local rebuild test, and also updated the patch according to Ken's review.

**Review by liw — APPROVED** [review]

File: src/container/srv_container.c [hunk 6]
```diff
@@ -1181,52 +1184,39 @@ cont_open(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl, struct cont *cont,
...
+		if (rc)
+			D_GOTO(out, rc);
 	}
 
-out_rpc:
-	crt_req_decref(rpc);
+	if (DAOS_FAIL_CHECK(DAOS_CONT_CLOSE_FAIL_CORPC))
+		rc = -DER_TIMEDOUT;
 out:
 	D_DEBUG(DF_DSMS, DF_CONT": bcasted: hdls[0]="DF_UUID" nhdls=%d: %d\n",
```
Comment (kccain) [inline src/container/srv_container.c:1221]: probably (in a future patch) want to change the text of this warning from bcasted -> closed

**Review by kccain — APPROVED** [review]

**Review by sydneyvanda — APPROVED** [review]

**Comment by johannlombardi** [comment]
Could you please have a look at the failure?

**Comment by wangdi1** [comment]
The failure caused by DAOS-5715 and DAOS-5807.

**Pull request merged**